### PR TITLE
Add support for kTLS zerocopy sendfile on Linux

### PR DESCRIPTION
**Pull request by nvmmax**
TLS device offload allows to perform zerocopy sendfile transmissions.
FreeBSD provides this feature by default, and Linux 5.19 introduced it
as an opt-in. Zerocopy improves the TX rate significantly, but has a
side effect: if the underlying file is changed while being transmitted,
and a TCP retransmission happens, the receiver may get a TLS record
containing both new and old data, which leads to an authentication
failure and termination of connection. This effect is the reason Linux
makes a copy on sendfile by default.

This commit adds support for TLS zerocopy sendfile on Linux and uses it
by default, aligning OpenSSL's behavior on Linux and FreeBSD. A new
option called KTLSTxNoZerocopySendfile is added to disable the new
behavior on Linux and revert to copying. This option should be used when
the application can't guarantee that the file is not modified, and it's
not desired to break the connection.

The related documentation is also added in this commit. The unit test is
not added, because testing this functionality requires specific hardware
and a non-local peer.

Signed-off-by: Maxim Mikityanskiy <maximmi@nvidia.com>
Reviewed-by: Tariq Toukan <tariqt@nvidia.com>
Reviewed-by: Boris Pismenny <borisp@nvidia.com>

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by nvmmax** [comment]
Thanks for reminding me, I'm aware of that, but unfortunately it takes time for the authorized people to sign the corporate CLA.

**Comment by nvmmax** [comment]
Can the CI be triggered for my pull request?

**Comment by mattcaswell** [comment]
> Can the CI be triggered for my pull request?

Done

**Comment by galpress** [comment]
Hey, can someone please take a look at this patch?

**Comment by ttoukan** [comment]
Hi, what's the status of this one? I see that all checks passed. Is there any blocker?

File: doc/man3/SSL_CONF_cmd.pod [hunk 1]
```diff
@@ -535,6 +535,13 @@ B<SSL_OP_ENABLE_KTLS>.
 B<StrictCertCheck>: Enable strict certificate checking. Equivalent to
 setting B<SSL_CERT_FLAG_TLS_STRICT> with SSL_CTX_set_cert_flags().
 
+B<KTLSTxNoZerocopySendfile>: disable the zerocopy TX mode of sendfile() on
+Linux. This mode is enabled by default on supported kernels and gives a
+performance boost when used with kTLS device offload. Note that invalid TLS
+records might be transmitted if the file is modified while being sent, so this
+option can be used to disable zerocopy. This option has no effect if B<KTLS> is
+not enabled. Equivalient to B<SSL_OP_DISABLE_KTLS_TX_ZEROCOPY_SENDFILE>.
```
Comment (hlandau) [inline doc/man3/SSL_CONF_cmd.pod:543]: typo: Equivalent

**Review by hlandau — CHANGES_REQUESTED** [review]
One nit, otherwise looks good.

File: doc/man3/SSL_CONF_cmd.pod (hunk 1 above)
Comment (nvmmax) [reply doc/man3/SSL_CONF_cmd.pod]: Fixed.

File: doc/man1/openssl-s_server.pod.in [hunk 2]
```diff
@@ -781,6 +782,14 @@ instead of BIO_write() to send the HTTP response requested by a client.
 This option is only valid when B<-ktls> along with B<-WWW> or B<-HTTP>
 are specified.
 
+=item B<-no_zerocopy_sendfile>
+
+By default, SSL_sendfile() will use the zerocopy TX mode, which gives a
+performance boost when used with kTLS device offload. Note that invalid TLS
+records might be transmitted if the file is modified while being sent, so this
+option can be used to disable zerocopy. This option is only valid when
+B<-sendfile> is specified.
```
Comment (mattcaswell) [inline doc/man1/openssl-s_server.pod.in:791]: It seems to me that by having this on by default you are introducing a breaking change. Stuff that worked before may no longer work. That's not allowed in a minor release, so it might be better to have it off by default and explicitly ask for it to be on.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Seems like this PR justifies an entry in CHANGES.md

File: doc/man1/openssl-s_server.pod.in (hunk 2 above)
Comment (nvmmax) [reply doc/man1/openssl-s_server.pod.in:791]: OpenSSL already supports zerocopy TLS sendfile on FreeBSD, and it's always active (i.e. there isn't even an option to disable it). The disclaimer that the file shouldn't be modified during sendfile already applies to FreeBSD to the same extent.

Since OpenSSL is a cross-platform library:

- The behavior on different platforms should be as close as possible, i.e. if OpenSSL on FreeBSD uses zerocopy by default, OpenSSL on Linux should use it too.
- I would assume that applications using a cross-platform library do not rely on implicit platform-specific behaviors, but stick to the strictest limitations among all platforms. As FreeBSD may corrupt TLS records if the application modifies the file during sendfile, applications using OpenSSL shouldn't modify the file if they care about this corruption. That means, it's safe to turn on the similar feature for Linux.

So, it's a breaking change only for the applications that are already broken (ones that modify the file during sendfile and don't want the corruption to happen, but when they are compiled on FreeBSD, they already suffer from it).

Having it off by default on Linux and always on on FreeBSD means introducing platform-specific behaviors exposed to the users of a cross-platform library, who expect it to behave the same way on all platforms.

Comment (mattcaswell) [reply doc/man1/openssl-s_server.pod.in:791]: > I would assume that applications using a cross-platform library do not rely on implicit platform-specific behaviors,

I don't think this is a safe assumption. Although OpenSSL is a cross-platform library, not all users of OpenSSL care about its cross-platform capabilities.

> As FreeBSD may corrupt TLS records if the application modifies the file during sendfile, applications using OpenSSL shouldn't modify the file if they care about this corruption. 

Users on Linux are unlikely to be aware of this if their application does not run on FreeBSD.

Comment (nvmmax) [reply doc/man1/openssl-s_server.pod.in:791]: > That's not allowed in a minor release

What about merging it to the next major release? Is that allowed?

Comment (mattcaswell) [reply doc/man1/openssl-s_server.pod.in:791]: > What about merging it to the next major release? Is that allowed?

In principle yes - but in practice there is no branch for it as yet. master is being used for 3.1.

We could choose to take this to OTC to determine whether this is sufficiently breaking to justify requiring that this is "off" by default.

**Comment by mattcaswell** [comment]
** OTC Question: Should zero-copy sendfile be the default on Linux? This is already the default on FreeBSD but isn't currently on Linux - so there is an inconsistency. Changing this could theoretically break some Linux users **

**Comment by MILASTARX84** [comment]
technically i have found out you can text your self thereof people cream
pretend to be someone only to be a third party pulling all the strings and
getting away with ruining peoples lives

**Comment by mattcaswell** [comment]
**OTC: We should not change the default for Linux in 3.x. We might reconsider for 4.0. This PR would be ok if it was an opt-in runtime configuration.**

**Comment by nvmmax** [comment]
I see an error in CI, but it doesn't seem related to my patch, since I didn't touch this file. Or am I missing something?

```
crypto/bn/rsaz_exp_x2.c:562:20: runtime error: load of misaligned address 0x612000000dc6 for type 'uint64_t', which requires 8 byte alignment
```

It wasn't there before the rebase.

**Comment by t8m** [comment]
One more rebase will be needed.

**Comment by nvmmax** [comment]
With the recent changes, I see that the relevant piece of code, which configured ktls, disappeared from ssl/tls13_enc.c. Please advise whether my adaptation in ssl/record/methods/ktls_meth.c is valid.

File: ssl/record/methods/ktls_meth.c [hunk 3]
```diff
@@ -423,6 +423,9 @@ static int ktls_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
     if (!BIO_set_ktls(rl->bio, &crypto_info, rl->direction))
         return OSSL_RECORD_RETURN_NON_FATAL_ERR;
 
+    if (rl->direction && (rl->options & SSL_OP_ENABLE_KTLS_TX_ZEROCOPY_SENDFILE))
+        BIO_set_ktls_tx_zerocopy_sendfile(rl->bio);
```
Comment (mattcaswell) [inline ssl/record/methods/ktls_meth.c:427]: Should we check the return value here?

File: ssl/t1_enc.c [hunk 4]
```diff
@@ -437,8 +437,12 @@ int tls1_change_cipher_state(SSL_CONNECTION *s, int which)
         goto skip_ktls;
 
     /* ktls works with user provided buffers directly */
-    if (BIO_set_ktls(bio, &crypto_info, which & SSL3_CC_WRITE))
+    if (BIO_set_ktls(bio, &crypto_info, which & SSL3_CC_WRITE)) {
+        if ((s->options & SSL_OP_ENABLE_KTLS_TX_ZEROCOPY_SENDFILE) &&
+            (which & SSL3_CC_WRITE))
+                BIO_set_ktls_tx_zerocopy_sendfile(bio);
         SSL_set_options(SSL_CONNECTION_GET_SSL(s), SSL_OP_NO_RENEGOTIATION);
+    }
```
Comment (mattcaswell) [inline ssl/t1_enc.c:445]: Side note: This is needed right now, but all this code will be deleted in the not-too-distant future as part of my write record layer refactor work. the changes in ktls_meth.c will be sufficient.

File: doc/man3/SSL_CTX_set_options.pod [hunk 5]
```diff
@@ -175,6 +175,13 @@ by the kernel directly and not via any available OpenSSL Providers. This might
 be undesirable if, for example, the application requires all cryptographic
 operations to be performed by the FIPS provider.
 
+=item SSL_OP_ENABLE_KTLS_TX_ZEROCOPY_SENDFILE
+
+With this option, sendfile() will use the zerocopy mode, which gives a
+performance boost when used with kTLS hardware offload. Note that invalid TLS
+records might be transmitted if the file is changed while being sent. This
+option has no effect if B<SSL_OP_ENABLE_KTLS> is not enabled.
```
Comment (mattcaswell) [inline doc/man3/SSL_CTX_set_options.pod:183]: Should we mention that some platforms may do this by default anyway?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
>  Please advise whether my adaptation in ssl/record/methods/ktls_meth.c is valid.

Looks fine aside from my minor comments above.

File: include/openssl/bio.h.in [hunk 6]
```diff
@@ -188,14 +188,23 @@ extern "C" {
...
+ * # define BIO_CTRL_GET_KTLS_TX_ZEROCOPY_SENDFILE 86
+ */
+
 # ifndef OPENSSL_NO_KTLS
 #  define BIO_get_ktls_send(b)         \
      (BIO_ctrl(b, BIO_CTRL_GET_KTLS_SEND, 0, NULL) > 0)
 #  define BIO_get_ktls_recv(b)         \
      (BIO_ctrl(b, BIO_CTRL_GET_KTLS_RECV, 0, NULL) > 0)
+#  define BIO_get_ktls_zerocopy_sendfile(b) \
+     (BIO_ctrl(b, BIO_CTRL_GET_KTLS_TX_ZEROCOPY_SENDFILE, 0, NULL) > 0)
```
[the remark this answers is not in the document]
Comment (nvmmax) [reply include/openssl/bio.h.in:203]: Removed altogether, as it's not used internally either.

File: ssl/record/methods/ktls_meth.c [hunk 7]
```diff
@@ -423,6 +423,9 @@ static int ktls_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
     if (!BIO_set_ktls(rl->bio, &crypto_info, rl->direction))
         return OSSL_RECORD_RETURN_NON_FATAL_ERR;
 
+    if (rl->direction && (rl->options & SSL_OP_ENABLE_KTLS_TX_ZEROCOPY_SENDFILE))
```
[the remark this answers is not in the document]
Comment (nvmmax) [reply ssl/record/methods/ktls_meth.c:426]: OK.

File: ssl/record/methods/ktls_meth.c (hunk 3 above)
Comment (nvmmax) [reply ssl/record/methods/ktls_meth.c:427]: No need, it's an opportunistic optimization, it may be unavailable on the running kernel, and the application opting in for it shouldn't get an error in such a case.

File: ssl/t1_enc.c (hunk 4 above)
Comment (nvmmax) [reply ssl/t1_enc.c:445]: OK, I'm keeping it for now, feel free to remove it when it becomes unneeded.

File: doc/man3/SSL_CTX_set_options.pod (hunk 5 above)
Comment (nvmmax) [reply doc/man3/SSL_CTX_set_options.pod:183]: OK.

File: crypto/bio/bss_sock.c [hunk 8]
```diff
@@ -235,6 +235,13 @@ static long sock_ctrl(BIO *b, int cmd, long num, void *ptr)
         BIO_clear_ktls_ctrl_msg_flag(b);
         ret = 0;
         break;
+    case BIO_CTRL_SET_KTLS_TX_ZEROCOPY_SENDFILE:
+        ret = ktls_enable_tx_zerocopy_sendfile(b->num);
+        if (ret)
+            BIO_set_ktls_zerocopy_sendfile_flag(b);
+        break;
+    case BIO_CTRL_GET_KTLS_TX_ZEROCOPY_SENDFILE:
```
Comment (mattcaswell) [inline crypto/bio/bss_sock.c:243]: Not called so remove?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/ssl_conf.c [hunk 9]
```diff
@@ -397,7 +397,8 @@ static int cmd_Options(SSL_CONF_CTX *cctx, const char *value)
         SSL_FLAG_TBL_INV("ExtendedMasterSecret", SSL_OP_NO_EXTENDED_MASTER_SECRET),
         SSL_FLAG_TBL_INV("CANames", SSL_OP_DISABLE_TLSEXT_CA_NAMES),
         SSL_FLAG_TBL("KTLS", SSL_OP_ENABLE_KTLS),
-        SSL_FLAG_TBL_CERT("StrictCertCheck", SSL_CERT_FLAG_TLS_STRICT)
+        SSL_FLAG_TBL_CERT("StrictCertCheck", SSL_CERT_FLAG_TLS_STRICT),
+        SSL_FLAG_TBL("KTLSTxZerocopySendfile", SSL_OP_ENABLE_KTLS_TX_ZEROCOPY_SENDFILE),
```
[the remark this answers is not in the document]
Comment (nvmmax) [reply ssl/ssl_conf.c:401]: There is no way to unit test it, because it requires special hardware (a NIC that supports kTLS device offload). I mentioned it in the commit message:

> The unit test is not added, because testing this functionality requires specific hardware and a non-local peer.

File: crypto/bio/bss_conn.c [hunk 10]
```diff
@@ -600,6 +600,13 @@ static long conn_ctrl(BIO *b, int cmd, long num, void *ptr)
         BIO_clear_ktls_ctrl_msg_flag(b);
         ret = 0;
         break;
+    case BIO_CTRL_SET_KTLS_TX_ZEROCOPY_SENDFILE:
+        ret = ktls_enable_tx_zerocopy_sendfile(b->num);
+        if (ret)
+            BIO_set_ktls_zerocopy_sendfile_flag(b);
+        break;
+    case BIO_CTRL_GET_KTLS_TX_ZEROCOPY_SENDFILE:
+        return BIO_should_ktls_zerocopy_sendfile_flag(b) != 0;
```
[the remark this answers is not in the document]
Comment (nvmmax) [reply crypto/bio/bss_conn.c:609]: OK.

File: crypto/bio/bss_sock.c (hunk 8 above)
Comment (nvmmax) [reply crypto/bio/bss_sock.c:243]: OK.

File: include/openssl/bio.h.in [hunk 11]
```diff
@@ -188,6 +188,12 @@ extern "C" {
 # define BIO_DGRAM_CAP_PROVIDES_SRC_ADDR    (1U << 2)
 # define BIO_DGRAM_CAP_PROVIDES_DST_ADDR    (1U << 3)
 
+/*
+ * internal BIO:
+ * # define BIO_CTRL_SET_KTLS_TX_ZEROCOPY_SENDFILE 85
+ * # define BIO_CTRL_GET_KTLS_TX_ZEROCOPY_SENDFILE 86
```
[the remark this answers is not in the document]
Comment (nvmmax) [reply include/openssl/bio.h.in:194]: OK.

File: ssl/ssl_conf.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (mattcaswell) [reply ssl/ssl_conf.c:401]: We can at least test that it is possible to set the option and it doesn't break anything if you do. Testing that it actually *does* anything is a lot harder - but I don't think we need to do that.

**Comment by wan7teen-ping** [comment]
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Jum, 24 Jun 2022, 5:28 PM nvmmax ***@***.***> menulis:

**Comment by MILASTARX84** [comment]
So much-needed to trust and to keep faith that ur not having sex with Paula

**Comment by MILASTARX84** [comment]
U get this

**Review by mattcaswell — APPROVED** [review]
LGTM. Although it looks like a rebase is required.

**Comment by mattcaswell** [comment]
Ping @openssl/committers for second review

File: apps/s_server.c [hunk 12]
```diff
@@ -1721,6 +1730,11 @@ int s_server_main(int argc, char *argv[])
 #endif
 
 #ifndef OPENSSL_NO_KTLS
+    if (use_zc_sendfile && !use_sendfile) {
+        BIO_printf(bio_out, "Warning: -zerocopy_sendfile depends on -sendfile, enabling -sendfile now.\n");
```
Comment (tmshort) [inline apps/s_server.c:1734]: This message is not really necessary.

File: doc/man1/openssl-s_server.pod.in [hunk 13]
```diff
@@ -781,6 +782,14 @@ instead of BIO_write() to send the HTTP response requested by a client.
 This option is only valid when B<-ktls> along with B<-WWW> or B<-HTTP>
 are specified.
 
+=item B<-zerocopy_sendfile>
+
+If this option is set, SSL_sendfile() will use the zerocopy TX mode, which gives
+a performance boost when used with kTLS hardware offload. Note that invalid
+TLS records might be transmitted if the file is changed while being sent.
+This option is only valid when B<-sendfile> is specified. Note that kTLS
```
Comment (tmshort) [inline doc/man1/openssl-s_server.pod.in:790]: This statement isn't true, as doing `zerocopy_sendfile` turns on `sendfile` automatically (although with a warning). I would instead get rid of that warning, and just state that `zerocopy_sendfile` implies the `sendfile` options and leave it at that.

File: doc/man3/SSL_CONF_cmd.pod [hunk 14]
```diff
@@ -535,6 +535,14 @@ B<SSL_OP_ENABLE_KTLS>.
 B<StrictCertCheck>: Enable strict certificate checking. Equivalent to
 setting B<SSL_CERT_FLAG_TLS_STRICT> with SSL_CTX_set_cert_flags().
 
+B<KTLSTxZerocopySendfile>: use the zerocopy TX mode of sendfile(), which gives
+a performance boost when used with kTLS hardware offload. Note that invalid TLS
+records might be transmitted if the file is changed while being sent. This
```
Comment (tmshort) [inline doc/man3/SSL_CONF_cmd.pod:540]: > Note that invalid TLS records might be transmitted if the file is changed while being sent.

This should be emphasized A LOT more. Big flags around it with blinking LEDS.

File: include/openssl/ssl.h.in [hunk 15]
```diff
@@ -410,6 +410,8 @@ typedef int (*SSL_async_callback_fn)(SSL *s, void *arg);
      * interoperability with CryptoPro CSP 3.x
      */
 # define SSL_OP_CRYPTOPRO_TLSEXT_BUG                     SSL_OP_BIT(31)
+    /* Enable kTLS TX zerocopy on Linux */
+# define SSL_OP_ENABLE_KTLS_TX_ZEROCOPY_SENDFILE         SSL_OP_BIT(32)
```
Comment (tmshort) [inline include/openssl/ssl.h.in:414]: Yay for 64bit options!

File: ssl/ssl_conf.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (tmshort) [reply ssl/ssl_conf.c:401]: Yeah, testing for TFO required looking at some behavior, but not erroring out, as long as data still passed.

**Review by tmshort — CHANGES_REQUESTED** [review]
Documentation needs updates

File: apps/s_server.c (hunk 12 above)
Comment (nvmmax) [reply apps/s_server.c:1734]: I added it to be consistent with:

```
    if (use_sendfile && enable_ktls == 0) {
        BIO_printf(bio_out, "Warning: -sendfile depends on -ktls, enabling -ktls now.\n");
        enable_ktls = 1;
    }
```

and with the documentation that says that `-zerocopy_sendfile` is only valid with `-sendfile` (similar to `-sendfile` and `-ktls`).

Do you suggest that this message isn't needed while the next one (`-sendfile` vs `-ktls`) is still needed? Why?

File: doc/man1/openssl-s_server.pod.in (hunk 13 above)
Comment (nvmmax) [reply doc/man1/openssl-s_server.pod.in:790]: It's the same as:

```
=item B<-sendfile>

If this option is set and KTLS is enabled, SSL_sendfile() will be used
instead of BIO_write() to send the HTTP response requested by a client.
This option is only valid when B<-ktls> along with B<-WWW> or B<-HTTP>
are specified.
```

I followed this existing example. If you think this pattern is not valid, I suggest fixing it outside the scope of this pull request.

File: doc/man3/SSL_CONF_cmd.pod (hunk 14 above)
Comment (nvmmax) [reply doc/man3/SSL_CONF_cmd.pod:540]: Any specific suggestions how to rephrase it?

Considering that `sendfile()` manpage explicitly prohibits modifying the file:

```
       If out_fd refers to a socket or pipe with zero-copy support,
       callers must ensure the transferred portions of the file referred
       to by in_fd remain unmodified until the reader on the other end
       of out_fd has consumed the transferred data.
```

and OpenSSL on FreeBSD has the same limitation, but it's even worse, because it is the default, can't be turned off and not documented anywhere in OpenSSL, I think the Linux side is documented good enough.

File: doc/man1/openssl-s_server.pod.in (hunk 13 above)
Comment (tmshort) [reply doc/man1/openssl-s_server.pod.in:790]: The problem I see is that you are saying an argument is _required_, yet it isn't. It's turned on automatically. In other words, the documentation says you have to specify:
```
openssl s_server -ktls -sendfile -zerocopy_sendfile
```
When in actuality, the following works.
```
openssl s_server -zerocopy_sendfiler
```
In this case, the `-sendfile` and `k'tls` options are enabled automatically and are _not_ required.

So, either the code needs to change to not automatically enable those options, or the documentation needs to change to state that `-ktls` and `-sendfile` are automatically enabled. Stating they may be required may be technically correct (the _best_ type of correct), but "required" has a specific meaning in terms of UI/UX.

File: doc/man3/SSL_CONF_cmd.pod (hunk 14 above)
Comment (tmshort) [reply doc/man3/SSL_CONF_cmd.pod:540]: The problem is that users aren't necessarily looking at Linux documentation while reading the OpenSSL documentation.  It should probably made its own paragraph, possibly in a WARNINGS section, or with the text "WARNING" preceding it. Please look in `doc/man3` for further examples.
```
If this option is set, SSL_sendfile() will use the zerocopy TX mode, which gives
a performance boost when used with kTLS hardware offload. 
This option automatically enables the B<-sendfile> option.
Note that kTLS sendfile on FreeBSD always runs in the zerocopy mode.

WARNING: Invalid TLS records might be transmitted if the file is changed while being sent.
Please refer to the L<sendfile(3)> documentation.
```
Or:
```
=head1 WARNINGS

Invalid TLS records might be transmitted if the file is changed while being sent.
Please refer to the L<sendfile(3)> documentation.

File: doc/man1/openssl-s_server.pod.in (hunk 13 above)
Comment (nvmmax) [reply doc/man1/openssl-s_server.pod.in:790]: > So, either the code needs to change to not automatically enable those options, or the documentation needs to change to state that -ktls and -sendfile are automatically enabled.

I totally agree, and it makes total sense, but fixing `-sendfile` vs `-ktls` is outside of scope of my pull request, and fixing just `-zerocopy_sendfile` vs `-sendfile` will be inconsistent with `-sendfile` vs `-ktls`. Although, if you insist on writing the matching documentation for the new code, rather than consistency with the existing pattern, I can update the docs for `-zerocopy_sendfile`, and leave fixing the docs for `-sendfile` to someone else.

Comment (tmshort) [reply doc/man1/openssl-s_server.pod.in:790]: At the very least, update the documentation to keep it correct with respect to the code. I actually don't think that it's too far out of scope to update the documentation for the `-sendfile`/`-ktls` option, since it's in the same file. But I would accept this PR without that change.

File: doc/man3/SSL_CONF_cmd.pod (hunk 14 above)
Comment (nvmmax) [reply doc/man3/SSL_CONF_cmd.pod:540]: This warning is not Linux-specific. FreeBSD suffers from the same limitation, moreover it can't even be disabled. If Linux users don't read Linux documentation, then FreeBSD users don't do it either, so OpenSSL should have an even more prominent warning for FreeBSD, shouldn't it?

That is, it doesn't make sense to me to add the Linux warning when there is no FreeBSD warning, and documenting FreeBSD limitations is outside the scope of my pull request. Besides, it might be possible to document it somewhere in a single place for both Linux and FreeBSD.

Comment (tmshort) [reply doc/man3/SSL_CONF_cmd.pod:540]: Understood. Note that my suggestions were not Linux nor FreeBSD specific.

File: include/openssl/ssl.h.in (hunk 15 above)
Comment (tmshort) [reply include/openssl/ssl.h.in:414]: And with the certificate compression code added, this needs to be bumped up to `SSL_OP_BIT(34)`.

**Comment by galpress** [comment]
@tmshort thanks for the review!
Is anything else needed to get this merged?

File: apps/s_server.c (hunk 12 above)
Comment (tmshort) [reply apps/s_server.c:1741]: Still not a fan of this message, but it won't block my approval.

File: doc/man3/SSL_CONF_cmd.pod (hunk 14 above)
Comment (tmshort) [reply doc/man3/SSL_CONF_cmd.pod:566]: This still hasn't been addressed. I will also note that the phrasing below "This option only makes on Linux", is probably better as "This option only applies to Linux."

**Review by tmshort — CHANGES_REQUESTED** [review]
Still not a fan of some of the documentation and s_server output.

I will say this, that just because it's done in a certain way in one place, doesn't mean it's necessarily the right thing to do, or the only way to do it. 

The documentation is technically incorrect because the options are not really "required" if the option is automatically turned on. If that documentation is given to QA/DevTest, they will write up a bug on it.

In other words, the documentation and the behavior/message don't really match, and I'd like to see one or the other fixed, even the existing code.

**Comment by tmshort** [comment]
In addition, @hlandau would have to approve, and @mattcaswell would have to re-approve, as changes were made after his initial approval.

**Comment by t8m** [comment]
> > And there's a CI failure, which may or may not be relevant.
> 
> I don't think the CI failure on macOS is relevant for a Linux-specific change.

The MacOS buildbot worker is broken currently. So that failure is not relevant.

File: include/openssl/bio.h.in [hunk 16]
```diff
@@ -188,6 +188,11 @@ extern "C" {
 # define BIO_DGRAM_CAP_PROVIDES_SRC_ADDR    (1U << 2)
 # define BIO_DGRAM_CAP_PROVIDES_DST_ADDR    (1U << 3)
 
+/*
+ * internal BIO:
+ * # define BIO_CTRL_SET_KTLS_TX_ZEROCOPY_SENDFILE 85
```
Comment (t8m) [inline include/openssl/bio.h.in:193]: Please note the value overlaps with BIO_CTRL_DGRAM_GET_EFFECTIVE_CAPS. Please change it. Also the comment should be placed above the BIO_DGRAM_CAP.... defines.

Also the comment on line 233 needs to be updated to reserve the internal BIO_FLAGS value added.

File: include/internal/bio.h [hunk 17]
```diff
@@ -43,16 +43,20 @@ int bread_conv(BIO *bio, char *data, size_t datal, size_t *read);
 # define BIO_CTRL_SET_KTLS                      72
 # define BIO_CTRL_SET_KTLS_TX_SEND_CTRL_MSG     74
 # define BIO_CTRL_CLEAR_KTLS_TX_CTRL_MSG        75
+# define BIO_CTRL_SET_KTLS_TX_ZEROCOPY_SENDFILE 85
```
Comment (t8m) [inline include/internal/bio.h:46]: The value needs to be changed.

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man1/openssl-s_server.pod.in [hunk 18]
```diff
@@ -792,6 +793,15 @@ instead of BIO_write() to send the HTTP response requested by a client.
 This option is only valid when B<-ktls> along with B<-WWW> or B<-HTTP>
 are specified.
 
+=item B<-zerocopy_sendfile>
+
+If this option is set, SSL_sendfile() will use the zerocopy TX mode, which gives
+a performance boost when used with kTLS hardware offload. Note that invalid
```
Comment (tmshort) [inline doc/man1/openssl-s_server.pod.in:799]: I am not seeing "kTLS" (note capitalization) anywhere else in the documentation, except this PR.

Please change here, and elsewhere, to "Kernel TLS" or "KTLS" (all caps).

File: doc/man1/openssl-s_server.pod.in (hunk 13 above)
Comment (tmshort) [reply doc/man1/openssl-s_server.pod.in]: It seems you understand where I'm coming from, so perhaps a later PR.

File: include/internal/bio.h [hunk 19]
```diff
@@ -43,16 +43,20 @@ int bread_conv(BIO *bio, char *data, size_t datal, size_t *read);
...
  * BIO_FLAGS_KTLS_TX means we are using ktls with this BIO for sending.
  * BIO_FLAGS_KTLS_TX_CTRL_MSG means we are about to send a ctrl message next.
  * BIO_FLAGS_KTLS_RX means we are using ktls with this BIO for receiving.
+ * BIO_FLAGS_KTLS_TX_ZEROCOPY_SENDFILE means we are using the zerocopy mode with
+ * this BIO for sending using sendfile.
  */
 # define BIO_FLAGS_KTLS_TX_CTRL_MSG 0x1000
 # define BIO_FLAGS_KTLS_RX          0x2000
 # define BIO_FLAGS_KTLS_TX          0x4000
+# define BIO_FLAGS_KTLS_TX_ZEROCOPY_SENDFILE 0x8000
```
Comment (t8m) [inline include/internal/bio.h:59]: This internal flag needs to be reflected (as comment) in the openssl/bio.h as well. Otherwise we risk reusing the flag value.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Review by tmshort — APPROVED** [review]
Fine by me
